### PR TITLE
Competition Wrapper

### PR DESCRIPTION
**Pull request by Parth097**
This pull request introduces a new `CompetitionWrapper` component to manage the display of competition-related content based on specific date conditions. The component is integrated into various parts of the competition pages to handle the display of content before and after the competition end date.

Key changes include:

### New Component:

* [`src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx`](diffhunk://#diff-0531ea82021cac74e809bfbfd635458105f110f256077540eee344d69ee0ac73R1-R25): Created a `CompetitionWrapper` component that conditionally renders children based on the current date relative to the competition end date and a period of four weeks after the end date.

### Component Integration:

* [`src/app/components/pages/IsaacCompetition/EntryForm/EntryFormHandler.tsx`](diffhunk://#diff-d761348580dc61f898f9a4a9f655705918bb78b637816f4abd4627770946e7b2R7-R11): Integrated `CompetitionWrapper` to conditionally display the competition entry form and a message indicating that entries are closed after the competition end date. [[1]](diffhunk://#diff-d761348580dc61f898f9a4a9f655705918bb78b637816f4abd4627770946e7b2R7-R11) [[2]](diffhunk://#diff-d761348580dc61f898f9a4a9f655705918bb78b637816f4abd4627770946e7b2L42-R68)
* [`src/app/components/pages/IsaacCompetition/HomepageHighlight/HomepageHighlight.tsx`](diffhunk://#diff-c5908ca77d40b31f738119679f65b6198c99706d32c6aaae7591285647b799ceR4-R8): Wrapped the homepage highlight content in the `CompetitionWrapper` to manage its display based on the competition end date. [[1]](diffhunk://#diff-c5908ca77d40b31f738119679f65b6198c99706d32c6aaae7591285647b799ceR4-R8) [[2]](diffhunk://#diff-c5908ca77d40b31f738119679f65b6198c99706d32c6aaae7591285647b799ceR27)

### CODE REVIEW

File: src/app/components/pages/IsaacCompetition/Tests/CompetitionWrapper.test.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    jest.setSystemTime(new Date("2025-05-01T12:00:00")); // After 4 weeks from the end date
+    const { queryByText } = render(
+      <CompetitionWrapper afterEndDateChildren={<div>Entries for this competition have now closed</div>}>
+        <div>Competition is open</div>
+      </CompetitionWrapper>,
+    );
+    expect(queryByText("Competition is open")).not.toBeInTheDocument();
+    expect(queryByText("Entries for this competition have now closed")).not.toBeInTheDocument();
+  });
+});
```
Comment (sxpydo) [inline src/app/components/pages/IsaacCompetition/Tests/CompetitionWrapper.test.tsx:45]: Really good job on including tests for different scenarios.
I would consider adding a test case for the exact end date to ensure boundary conditions are handled correctly.

Otherwise, great work! 👏🏾

File: src/app/components/pages/IsaacCompetition/HomepageHighlight/HomepageHighlight.tsx [hunk 2]
```diff
@@ -1,27 +1,30 @@
...
-            ]}
-          />
-        </Col>
-      </Row>
-    </Container>
+    <CompetitionWrapper>
+      <Container className="pt-2 pb-5">
+        <Row className="homepage-highlight rounded justify-content-center">
+          <Col xs={12} className="text-center">
+            <h1 className="homepage-highlight-sub-title py-4">2025 National Computer Science competition</h1>
```
Comment (sxpydo) [inline src/app/components/pages/IsaacCompetition/HomepageHighlight/HomepageHighlight.tsx:12]: For easier maintenance, try extracting the competition title and subtitle into constants or props. 😃

File: src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+interface CompetitionWrapperProps {
+  children: React.ReactNode;
+  afterEndDateChildren?: React.ReactNode;
+}
+
+const CompetitionWrapper = ({ children, afterEndDateChildren }: CompetitionWrapperProps) => {
+  const currentDate = new Date();
+
+  if (currentDate > endDate && currentDate <= fourWeeksAfterEndDate) {
+    return <>{afterEndDateChildren}</>;
```
Comment (sxpydo) [inline src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx:15]: Use a more descriptive name for the `afterEndDateChildren` prop, such as `closedCompetitionContent` or something simliar.

File: src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
+
+const endDate = new Date("2025-03-28T16:00:00"); // 4 PM on Friday, March 28, 2025
+const fourWeeksAfterEndDate = new Date(endDate.getTime() + 4 * 7 * 24 * 60 * 60 * 1000); // 4 weeks after end date
+
```
Comment (sxpydo) [inline src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx:5]: Could we extract the date calculations into separate constants or utility functions for better readability

File: src/app/components/pages/IsaacCompetition/Tests/CompetitionWrapper.test.tsx (hunk 1 above)
Comment (Parth097) [reply src/app/components/pages/IsaacCompetition/Tests/CompetitionWrapper.test.tsx:45]: These test dates are exactly on the competition end date.
It follow from end date, within 4 weeks & after 4 weeks of the end date.

File: src/app/components/pages/IsaacCompetition/HomepageHighlight/HomepageHighlight.tsx (hunk 2 above)
Comment (Parth097) [reply src/app/components/pages/IsaacCompetition/HomepageHighlight/HomepageHighlight.tsx:12]: Homepage highlight is not part of this work. Just the wrapper has been added around it. We can work on this after this ticket for clean up.

File: src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx (hunk 3 above)
Comment (Parth097) [reply src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx:15]: Good idea. I will change it to the suggested name

File: src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx (hunk 4 above)
Comment (Parth097) [reply src/app/components/pages/IsaacCompetition/CompetitionWrapper.tsx:5]: Added dateUtils.ts file to store the dates and calculations

**Review by sxpydo — APPROVED** [review]
